pkg/storage/batcheval/cmd_add_sstable_test.go

```diff
@@ -383,3 +383,146 @@ func TestAddSSTableMVCCStats(t *testing.T) {
 		t.Errorf("mvcc stats mismatch: diff(expected, actual): %s", pretty.Diff(afterStats, evaledStats))
 	}
 }
+
+func TestAddSSTableDisallowShadowing(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	e := engine.NewInMem(roachpb.Attributes{}, 1<<20)
+	defer e.Close()
+
+	for _, kv := range mvccKVsFromStrs([]strKv{
+		{"a", 1, "aa"},
+		{"a", 6, ""},
+		{"g", 5, "bb"},
+		{"r", 1, "rr"},
+		{"z", 2, "zz"},
+	}) {
+		if err := e.Put(kv.Key, kv.Value); err != nil {
+			t.Fatalf("%+v", err)
+		}
+	}
+
+	getSSTBytes := func(sstKVs []engine.MVCCKeyValue) []byte {
+		sst, err := engine.MakeRocksDBSstFileWriter()
+		if err != nil {
+			t.Fatalf("%+v", err)
+		}
+		defer sst.Close()
+		for _, kv := range sstKVs {
+			if err := sst.Add(kv); err != nil {
+				t.Fatalf("%+v", err)
+			}
+		}
+		sstBytes, err := sst.Finish()
+		if err != nil {
+			t.Fatalf("%+v", err)
+		}
+		return sstBytes
+	}
+
+	// Test key collision when ingesting an existing key with a lower ts.
```

Comment:
the interesting cases I can think of are all the combinations of: 
collision is first key, last key, some middle key of sst or of existing.
collision key is/is not equal to the header start key.
header start and end are existing keys.
higher ts is on sst or existing.
collision key is deleted, deleted and the re-added.
collision key is an intent.

I can see some of these are already covered but I'd either make a big comment up front explaining all the cases and then which test covers them, or make the comments on each test more verbose.